ddsrouter_core/src/cpp/writer/implementations/rtps/rpc/RequestWriter.cpp

```diff
@@ -0,0 +1,111 @@
+#include <fastrtps/rtps/participant/RTPSParticipant.h>
+#include <fastrtps/rtps/common/CacheChange.h>
+
+#include <writer/implementations/rtps/rpc/RequestWriter.hpp>
+#include <ddsrouter_utils/Log.hpp>
+
+namespace eprosima {
+namespace ddsrouter {
+namespace core {
+namespace rtps {
+
+using namespace eprosima::ddsrouter::core::types;
+
+RequestWriter::RequestWriter(
+        const ParticipantId& participant_id,
+        const RPCTopic& topic,
+        std::shared_ptr<PayloadPool> payload_pool,
+        fastrtps::rtps::RTPSParticipant* rtps_participant,
+        std::shared_ptr<ServiceRegistry> service_registry)
+    : Writer(participant_id, topic.request_topic(), payload_pool, rtps_participant)
+    , service_registry_(service_registry)
+{
+    logInfo(DDSROUTER_RTPS_REQUESTWRITER, "New RequestWriter created for topic " << topic_);
+}
+
+RequestWriter::~RequestWriter()
+{
+    logInfo(DDSROUTER_RTPS_REQUESTWRITER, "Deleting RequestWriter for topic " << topic_);
+}
+
+utils::ReturnCode RequestWriter::write_(
+        std::unique_ptr<DataReceived>& data) noexcept
+{
+    // if (participant_id_ != service_registry_->server_participant_id())
+    // {
+    //     // send request only when there is a server to receive it
+    //     // change return code if filtering logic is moved to Track
+    //     return utils::ReturnCode::RETCODE_OK;
+    // }
+
+    // Take new Change from history
+    fastrtps::rtps::CacheChange_t* new_change = rtps_writer_->new_change(eprosima::fastrtps::rtps::ChangeKind_t::ALIVE);
+
+    // If still is not able to get a change, return an error code
+    if (!new_change)
+    {
+        return utils::ReturnCode::RETCODE_ERROR;
+    }
+
+    // Get the Payload (copying it)
+    eprosima::fastrtps::rtps::IPayloadPool* payload_owner = payload_pool_.get();
+    if (!payload_pool_->get_payload(data->payload, payload_owner, (*new_change)))
+    {
+        logDevError(DDSROUTER_RTPS_REQUESTWRITER, "Error getting Payload.");
+        return utils::ReturnCode::RETCODE_ERROR;
+    }
+
+    logDebug(DDSROUTER_RTPS_REQUESTWRITER,
+            "Writer " << *this << " sending payload " << new_change->serializedPayload << " from " <<
+            data->source_guid);
+
+    eprosima::fastrtps::rtps::WriteParams wparams = new_change->write_params;
+    SampleIdentity reply_related_sample_identity = data->write_params.sample_identity();
+    reply_related_sample_identity.sequence_number(data->sequenceNumber);
+    // if (reply_related_sample_identity == SampleIdentity::Unknown())
+    // {
+    //     logWarning();
+        // return utils::ReturnCode::RETCODE_ERROR;
+    // }
+
+    // Thread safe as registry's related_sample_identity is set only in creation, when this writer is still disabled
+    // Also note that concurrent \c write_ operations in the same \c RequestWriter are not possible, as \c write from parent \c BaseWriter is guarded
+    wparams.related_sample_identity(service_registry_->related_sample_identity_nts());
```

Comment:
Could not this be done by every Writer?
Doing so and adding to service_registry when calling `write` and not here will avoid creating and maintaining this new class